specs/dynamicdependencies/DynamicDependencies.md

```diff
@@ -1281,6 +1281,101 @@ STDAPI MddBootstrapInitialize(
 /// @warning Packages made available via MddBootstrapInitialize() and
 ///          the Dynamic Dependencies API should not be used after this call.
 STDAPI_(void) MddBootstrapShutdown() noexcept;
+
+// C++ friendly APIs
+#if defined(__cplusplus)
+#if defined(WINDOWSAPPSDK_RELEASE_MAJORMINOR) && defined(WINDOWSAPPSDK_RELEASE_VERSION_TAG_W) && defined(WINDOWSAPPSDK_RUNTIME_VERSION_UINT64)
+namespace Microsoft::Windows::ApplicationModel
+{
+class PackageVersion : public PACKAGE_VERSION
+{
+public:
+    PackageVersion();
+
+    // Create an instance with the value `major.minor.build.revision`.
+    PackageVersion(uint16_t major, uint16_t minor = 0, uint16_t build = 0, uint16_t revision = 0);
+
+    // Create an instance from a version as a uint64.
+    PackageVersion(uint64_t version);
+
+    // Return the version as a uint64.
+    uint64_t ToVersion() const;
+
+#if defined(_XSTRING_) && defined(_STRSAFE_H_INCLUDED_) && defined(WI_VERIFY)
+    // Return the string as a formatted value "major.minor.build.revision".
+    std::wstring ToString() const;
+
+    static std::wstring ToString(uint16_t major, uint16_t minor, uint16_t build, uint16_t revision);
+#endif
+};
+
+namespace DynamicDependency::Bootstrap
+{
+    // Automate Boostrap shutdown when leaving scope
+    using unique_mddbootstrapshutdown_call = wil::unique_call<decltype(&::MddBootstrapShutdown), ::MddBootstrapShutdown>;
+
+    /// Call MddBootstrapInitialize and fail-fasts if it fails; returns an RAII objecct that reverts
+    ///
+    /// Initialize the calling process to use Windows App SDK's framework package.
+    ///
+    /// Find a Windows App SDK framework package meeting the criteria and make it available
+    /// for use by the current process. If multiple packages meet the criteria the best
+    /// candidate is selected.
+    ///
+    /// @param majorMinorVersion major and minor version of Windows App SDK's framework package, encoded as `0xMMMMNNNN` where M=Major, N=Minor (e.g. 1.2 == 0x00010002).
+    /// @param versionTag version tag (if any), e.g. "preview1".
+    /// @param minVersion the minimum version to use
+    /// @see Initialize(uint32_t, PCWSTR, PackageVersion)
+    /// @see Initialize_nothrow(uint32_t, PCWSTR, PackageVersion)
+    /// @see Shutdown()
+    WI_NODISCARD inline unique_mddbootstrapshutdown_call Initialize_failfast(
+        uint32_t majorMinorVersion = WINDOWSAPPSDK_RELEASE_MAJORMINOR,
+        PCWSTR versionTag = WINDOWSAPPSDK_RELEASE_VERSION_TAG_W,
+        PackageVersion minVersion = WINDOWSAPPSDK_RUNTIME_VERSION_UINT64);
+
+#if defined(WIL_ENABLE_EXCEPTIONS)
+    /// Call MddBootstrapInitialize and throws an exception if it fails; returns an RAII objecct that reverts
```

Comment:
```suggestion
    /// Call MddBootstrapInitialize and throws a wil::ResultException exception if it fails; returns an RAII object that reverts
```
Should clarify what kind of exception, so people can catch it.

Reply:
Should we? Or is that an implementation detail?

If we document it we're declaring the exception as part of the interface. Changing that in the future would be a breaking change.

C++ coding guidelines say (E.30: Don’t use exception specifications](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Re-specifications). Are you saying we should document exceptions but not declare them?

Do we have precedent or is this new territory we should bring up at API Review?

Reply:
Document the exceptions, but don't put them in the function prototype. (Starting in C++17, if you try to put them in the prototype, you get a compiler error.)

This is consistent with the C++ standard library, which documents what exceptions can be thrown by each method in the specification, but does not use `throw(blah)` annotations on the prototype.

Reply:
Per API Design:

* DO use C++/WinRT for error handling and exceptions/throwing
* DO NOT require WIL to use C++ APIs
* DO use std::unique_ptr<> with custom deleter for the cleanup method type (instead of wil::unique_call)
* DO document throwing a C++/WinRT exception (not just a std::exception)
  * the thrown C++/WinRT exception is part of the API contract. It would be a Breaking Change if it changes in the future